src/rgw/rgw_op.cc

```diff
@@ -937,36 +937,55 @@ void handle_replication_status_header(
 }
 
 /*
- * GET on CloudTiered objects is processed only when sent from the sync client.
- * In all other cases, fail with `ERR_INVALID_OBJECT_STATE`.
+ * GET on CloudTiered objects either it will synced to other zones.
+ * In all other cases, it will try to fetch the object from remote cloud endpoint.
  */
-int handle_cloudtier_obj(rgw::sal::Attrs& attrs, bool sync_cloudtiered) {
+int handle_cloudtier_obj(req_state* s, const DoutPrefixProvider *dpp, rgw::sal::Driver* driver, std::optional<uint64_t> days,
+                         rgw::sal::Attrs& attrs, bool sync_cloudtiered, RGWObjTier tier_config, optional_yield y)
+{
   int op_ret = 0;
-  auto attr_iter = attrs.find(RGW_ATTR_MANIFEST);
+  ldpp_dout(dpp, 20) << "reached handle cloud tier " << dendl;
+  if (sync_cloudtiered) {
+    bufferlist t, t_tier;
+    t.append("cloud-s3");
+    attrs[RGW_ATTR_CLOUD_TIER_TYPE] = t;
+    encode(tier_config, t_tier);
+    attrs[RGW_ATTR_CLOUD_TIER_CONFIG] = t_tier;
+	  return op_ret;
+  }
+
+  rgw::sal::Bucket* pbucket = NULL;
+  pbucket = s->bucket.get();
+  std::unique_ptr<rgw::sal::PlacementTier> tier;
+  rgw_placement_rule target_placement;
+
+  target_placement.inherit_from(pbucket->get_placement_rule());
+  auto attr_iter = attrs.find(RGW_ATTR_STORAGE_CLASS);
   if (attr_iter != attrs.end()) {
-    RGWObjManifest m;
-    try {
-      decode(m, attr_iter->second);
-      if (m.get_tier_type() == "cloud-s3") {
-        if (!sync_cloudtiered) {
-          /* XXX: Instead send presigned redirect or read-through */
-          op_ret = -ERR_INVALID_OBJECT_STATE;
-        } else { // fetch object for sync and set cloud_tier attrs
-          bufferlist t, t_tier;
-          RGWObjTier tier_config;
-          m.get_tier_config(&tier_config);
-
-          t.append("cloud-s3");
-          attrs[RGW_ATTR_CLOUD_TIER_TYPE] = t;
-          encode(tier_config, t_tier);
-          attrs[RGW_ATTR_CLOUD_TIER_CONFIG] = t_tier;
-        }
-      }
-    } catch (const buffer::end_of_buffer&) {
-      // ignore empty manifest; it's not cloud-tiered
-    } catch (const std::exception& e) {
-    }
+      target_placement.storage_class = attr_iter->second.to_str();
   }
+  op_ret = driver->get_zone()->get_zonegroup().get_placement_tier(target_placement, &tier);
+  ldpp_dout(dpp, 20) << "getting tier placement handle cloud tier" << op_ret <<
+                     " storage class " << target_placement.storage_class << dendl;
+  if (op_ret < 0) {
+   return op_ret;
+  }
+  rgw_bucket_dir_entry ent;
+  ent.key.name = s->object->get_key().name;
+  ent.meta.accounted_size = ent.meta.size = s->obj_size;
```

Comment:
I am not sure if this is right way to fill in ent. But probably instead of ent, we can directly pass on mtime/other variables actually needed in restore_obj* API. We can address it later.